cmd/pipeline-controller/reconciler.go

```diff
@@ -0,0 +1,220 @@
+		},
+	}
+	if err := builder.
+		ControllerManagedBy(mgr).
+		Named("pipeline-controller").
+		For(&v1.ProwJob{}).
+		WithOptions(controller.Options{MaxConcurrentReconciles: 5}).
+		Complete(reconciler); err != nil {
+		return nil, fmt.Errorf("failed to construct controller: %w", err)
+	}
+	return reconciler, nil
+}
+
+func (r *reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	log := r.logger.WithField("key", req.String()).WithField("prowjob", req.Name)
+	err := r.reconcile(ctx, req)
+	if err != nil {
+		log.WithError(err).Error("reconciliation failed")
+	}
+	return reconcile.Result{}, err
+}
+
+func (r *reconciler) cleanOldIds(interval time.Duration) {
+	ticker := time.NewTicker(interval)
+	defer ticker.Stop()
+
+	for range ticker.C {
+		r.ids.Range(func(key, value interface{}) bool {
+			if time.Since(value.(time.Time)) >= interval {
+				r.ids.Delete(key)
+			}
+			return true
+		})
+	}
+}
+
+func (r *reconciler) reconcile(ctx context.Context, req reconcile.Request) error {
+	var pj v1.ProwJob
+	if err := r.pjclientset.Get(ctx, req.NamespacedName, &pj); err != nil {
+		if errors.IsNotFound(err) {
+			return nil
+		}
+
+		return fmt.Errorf("failed to get prowjob %s: %w", req.String(), err)
+	}
+
+	if pj.Spec.Refs == nil || pj.Spec.Type != v1.PresubmitJob {
+		return nil
+	}
+
+	presubmits := r.configDataProvider.GetPresubmits(pj.Spec.Refs.Org + "/" + pj.Spec.Refs.Repo)
+	if len(presubmits.protected) == 0 && len(presubmits.alwaysRequired) == 0 && len(presubmits.conditionallyRequired) == 0 {
+		return nil
+	}
+
+	if status, err := r.reportSuccessOnPR(ctx, &pj, presubmits); err != nil || !status {
+		return err
+	}
+	if err := r.ghc.CreateComment(pj.Spec.Refs.Org, pj.Spec.Refs.Repo, pj.Spec.Refs.Pulls[0].Number, "/test remaining-required"); err != nil {
+		r.ids.Delete(composeKey(pj.Spec.Refs))
+		return err
+	}
+	return nil
+}
+
+func (r *reconciler) reportSuccessOnPR(ctx context.Context, pj *v1.ProwJob, presubmits presubmitTests) (bool, error) {
+	if pj == nil || pj.Spec.Refs == nil || len(pj.Spec.Refs.Pulls) != 1 {
+		return false, nil
+	}
+	selector := map[string]string{}
+	for _, l := range []string{kube.OrgLabel, kube.RepoLabel, kube.PullLabel, kube.BaseRefLabel} {
+		selector[l] = pj.ObjectMeta.Labels[l]
+	}
+	var pjs v1.ProwJobList
+	if err := r.lister.List(ctx, &pjs, ctrlruntimeclient.MatchingLabels(selector)); err != nil {
+		return false, fmt.Errorf("cannot list prowjob using selector %v", selector)
+	}
+
+	latestBatch := make(map[string]v1.ProwJob)
+	for _, pjob := range pjs.Items {
+		if pjob.Spec.Refs.Pulls[0].SHA == pj.Spec.Refs.Pulls[0].SHA {
+			if existing, ok := latestBatch[pjob.Spec.Job]; !ok {
+				latestBatch[pjob.Spec.Job] = pjob
+			} else if pjob.CreationTimestamp.After(existing.CreationTimestamp.Time) {
+				latestBatch[pjob.Spec.Job] = pjob
+			}
+		}
+	}
+
+	for _, presubmit := range presubmits.protected {
+		if _, ok := latestBatch[presubmit]; ok {
```

Comment:
Does it mean that if I have already manually triggered one of the jobs this controller would have triggered then it won't do anything?
As an example: let's suppose `e2e-1` and `e2e-2` on a repository, both being `protected`. Then I open a PR and I manually `/test e2e-1` but  no `e2e-2`. What does happen next?

Reply:
yes, if you trigger manually one of the jobs this controller the given SHA won't do anything. I think this is kind of logical. If user wants to speed-up execution then he/she takes the responsibility for the pipeline. It will be part of documentation.